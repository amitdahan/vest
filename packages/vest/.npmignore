# Autogenerated section. Do not edit manually.
node_modules
src
!types/
!dist/
tsconfig.json
!promisify/
!parser/
!enforce@schema/
!enforce@isURL/
!enforce@email/
!enforce@date/
!enforce@compounds/
!enforce@compose/
!debounce/
!classnames/
!SuiteSerializer/

# Manual Section. Edit at will.