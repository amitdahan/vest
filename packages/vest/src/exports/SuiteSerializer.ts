import { assign } from 'vest-utils';
import { IsolateSerializer } from 'vestjs-runtime';

import { CommonStates } from 'CommonStateMachine';
import { TIsolateSuite } from 'IsolateSuite';
import { IsolateTestPayload } from 'IsolateTest';
import { TestStatus } from 'IsolateTestStateMachine';
import { TFieldName, TGroupName } from 'SuiteResultTypes';
import { Suite } from 'SuiteTypes';
import { VestIsolateType } from 'VestIsolateType';
import { IsolateFocusedPayload } from 'focused';

export class SuiteSerializer {
  static serialize(suite: Suite<TFieldName, TGroupName>) {
    const dump = { ...suite.dump(), output: undefined };

    return IsolateSerializer.serialize(dump, MiniMap);
  }

  static deserialize(
    serialized: string | TIsolateSuite | Record<string, any>
  ): TIsolateSuite {
    return IsolateSerializer.deserialize(serialized, MiniMap) as TIsolateSuite;
  }

  static resume(
    suite: Suite<TFieldName, TGroupName>,
    root: string | TIsolateSuite | Record<string, any>
  ): void {
    const suiteRoot = SuiteSerializer.deserialize(root);

    suite.resume(suiteRoot);
  }
}

const testMiniMap: Record<keyof IsolateTestPayload, string> = {
  asyncTest: '_at', // asyncTest is not serialized
  fieldName: 'fN',
  groupName: 'gN',
  message: 'ms',
  severity: 'sv',
  status: 'st',
  testFn: '_tf', // testFn is not serialized
};

const focusMiniMap: Record<keyof IsolateFocusedPayload, string> = {
  focusMode: 'fM',
  match: 'm',
  matchAll: 'mA',
};

const MiniMap: MiniMap = {
  keys: {
    data: assign({}, testMiniMap, focusMiniMap),
  },
  values: {
    status: {
      CANCELED: 'C',
      DONE: 'D',
      FAILED: 'F',
      INITIAL: 'I',
      OMITTED: 'O',
      PASSING: 'P',
      PENDING: 'PE',
      SKIPPED: 'S',
      UNTESTED: 'U',
      WARNING: 'W',
    },
    $type: {
      Each: 'E',
      Focused: 'F',
      Group: 'G',
      OmitWhen: 'OW',
      SkipWhen: 'SW',
      Suite: 'S',
      Test: 'T',
    },
  },
};

type MiniMap = {
  keys: {
    data: Record<keyof IsolateTestPayload, string> &
      Record<keyof IsolateFocusedPayload, string>;
  };
  values: {
    status: Record<keyof typeof TestStatus, string> &
      Record<keyof typeof CommonStates, string>;
    $type: Record<keyof typeof VestIsolateType, string>;
  };
};
