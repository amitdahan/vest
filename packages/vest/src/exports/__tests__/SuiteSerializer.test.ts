import { SuiteSerializer } from 'SuiteSerializer';
import * as vest from 'vest';

describe('SuiteSerializer', () => {
  it('Should produce a valid serialized dump', () => {
    const suite = vest.create('suite_serialize_test', () => {
      vest.only('field_1');

      vest.test('field_1', 'field_1_message', () => false);
      vest.test('field_2', 'field_2_message', () => false);

      vest.group('group_1', () => {
        vest.test('field_3', 'field_3_message_1', () => false);
        vest.test('field_3', 'field_3_message_2', () => false);
        vest.test('field_4', 'field_4_message', () => false);
      });

      vest.skipWhen(false, () => {
        vest.test('field_5', 'field_5_message', () => false);
      });
    });
    suite();

    const serialized = SuiteSerializer.serialize(suite);
    expect(serialized).toMatchSnapshot();
  });
});

it('Should minify test payload', () => {
  const suite = vest.create('suite_serialize_test', () => {
    vest.test('field_1', 'field_1_message', () => false);
  });

  suite();
  const serialized = SuiteSerializer.serialize(suite);

  const parsed = JSON.parse(serialized);

  expect(parsed['C'][0]['D']).toBeDefined();
  expect(parsed['C'][0]['D']['fN']).toBe('field_1');
  expect(parsed['C'][0]['D']['msg']).toBe('field_1_message');
  expect(parsed['C'][0]['D']['sv']).toBe('error');
});

describe('suite.resume', () => {
  it('Should resume a suite from a serialized dump', () => {
    const suite = vest.create(() => {
      vest.only('field_1');

      vest.test('field_1', 'field_1_message', () => false);
      vest.test('field_2', 'field_2_message', () => false);

      vest.group('group_1', () => {
        vest.test('field_3', 'field_3_message_1', () => false);
        vest.test('field_3', 'field_3_message_2', () => false);
        vest.test('field_4', 'field_4_message', () => false);
      });

      vest.skipWhen(false, () => {
        vest.test('field_5', 'field_5_message', () => false);
      });
    });

    suite();
    suite.get();

    const serialized = SuiteSerializer.serialize(suite);

    const suite2 = vest.create(() => {});

    suite2();

    expect(suite.get()).not.toEqual(suite2.get());

    SuiteSerializer.resume(suite2, serialized);
    expect(suite.get()).isDeepCopyOf(suite2.get());

    expect(suite2.hasErrors()).toBe(true);
    expect(suite2.hasWarnings()).toBe(false);
    expect(suite2.get().tests.field_1).toBeDefined();

    suite2();
    expect(suite.get()).not.toEqual(suite2.get());
    expect(suite2.hasErrors()).toBe(false);
    expect(suite2.hasWarnings()).toBe(false);
    expect(suite2.get().tests.field_1).toBeUndefined();
  });
});
