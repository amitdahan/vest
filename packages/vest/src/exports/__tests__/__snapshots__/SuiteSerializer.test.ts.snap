// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuiteSerializer Should produce a valid serialized dump 1`] = `"{"C":[{"$":"Focused","D":{"focusMode":0,"match":["field_1"],"matchAll":false},"S":"DONE"},{"$":"Test","D":{"severity":"error","fieldName":"field_1","message":"field_1_message"},"S":"FAILED"},{"$":"Test","D":{"severity":"error","fieldName":"field_2","message":"field_2_message"},"S":"SKIPPED"},{"C":[{"$":"Test","D":{"severity":"error","fieldName":"field_3","groupName":"group_1","message":"field_3_message_1"},"S":"SKIPPED"},{"$":"Test","D":{"severity":"error","fieldName":"field_3","groupName":"group_1","message":"field_3_message_2"},"S":"SKIPPED"},{"$":"Test","D":{"severity":"error","fieldName":"field_4","groupName":"group_1","message":"field_4_message"},"S":"SKIPPED"}],"$":"Group","D":{},"S":"DONE"},{"C":[{"$":"Test","D":{"severity":"error","fieldName":"field_5","message":"field_5_message"},"S":"SKIPPED"}],"$":"SkipWhen","D":{},"S":"DONE"}],"$":"Suite","D":{"optional":{}},"S":"DONE"}"`;
